core/types/celo_block.go

```diff
@@ -0,0 +1,153 @@
+)
+
+type IstanbulExtra rlp.RawValue
+
+type beforeGingerbreadHeader struct {
+	ParentHash  common.Hash    `json:"parentHash"       gencodec:"required"`
+	Coinbase    common.Address `json:"miner"            gencodec:"required"`
+	Root        common.Hash    `json:"stateRoot"        gencodec:"required"`
+	TxHash      common.Hash    `json:"transactionsRoot" gencodec:"required"`
+	ReceiptHash common.Hash    `json:"receiptsRoot"     gencodec:"required"`
+	Bloom       Bloom          `json:"logsBloom"        gencodec:"required"`
+	Number      *big.Int       `json:"number"           gencodec:"required"`
+	GasUsed     uint64         `json:"gasUsed"          gencodec:"required"`
+	Time        uint64         `json:"timestamp"        gencodec:"required"`
+	Extra       []byte         `json:"extraData"        gencodec:"required"`
+}
+
+// This type is required to avoid an infinite loop when decoding
+type afterGingerbreadHeader Header
+
+func (h *Header) DecodeRLP(s *rlp.Stream) error {
+	var raw rlp.RawValue
+	err := s.Decode(&raw)
+	if err != nil {
+		return err
+	}
+
+	preGingerbread, err := isPreGingerbreadHeader(raw)
+	if err != nil {
+		return err
+	}
+
+	if preGingerbread { // Address
+		// Before gingerbread
+		decodedHeader := beforeGingerbreadHeader{}
+		err = rlp.DecodeBytes(raw, &decodedHeader)
+
+		h.ParentHash = decodedHeader.ParentHash
+		h.Coinbase = decodedHeader.Coinbase
+		h.Root = decodedHeader.Root
+		h.TxHash = decodedHeader.TxHash
+		h.ReceiptHash = decodedHeader.ReceiptHash
+		h.Bloom = decodedHeader.Bloom
+		h.Number = decodedHeader.Number
+		h.GasUsed = decodedHeader.GasUsed
+		h.Time = decodedHeader.Time
+		h.Extra = decodedHeader.Extra
+	} else {
+		// After gingerbread
+		decodedHeader := afterGingerbreadHeader{}
+		err = rlp.DecodeBytes(raw, &decodedHeader)
+
+		h.ParentHash = decodedHeader.ParentHash
+		h.UncleHash = decodedHeader.UncleHash
+		h.Coinbase = decodedHeader.Coinbase
+		h.Root = decodedHeader.Root
+		h.TxHash = decodedHeader.TxHash
+		h.ReceiptHash = decodedHeader.ReceiptHash
+		h.Bloom = decodedHeader.Bloom
+		h.Difficulty = decodedHeader.Difficulty
+		h.Number = decodedHeader.Number
+		h.GasLimit = decodedHeader.GasLimit
+		h.GasUsed = decodedHeader.GasUsed
+		h.Time = decodedHeader.Time
+		h.Extra = decodedHeader.Extra
+		h.MixDigest = decodedHeader.MixDigest
+		h.Nonce = decodedHeader.Nonce
+		h.BaseFee = decodedHeader.BaseFee
+		h.WithdrawalsHash = decodedHeader.WithdrawalsHash
+		h.BlobGasUsed = decodedHeader.BlobGasUsed
+		h.ExcessBlobGas = decodedHeader.ExcessBlobGas
+		h.ParentBeaconRoot = decodedHeader.ParentBeaconRoot
+	}
+
+	return err
+}
+
+// EncodeRLP implements encodes the Header to an RLP data stream.
+func (h *Header) EncodeRLP(w io.Writer) error {
+	// We check for a pre gingerbread header by looking for (GasLimit == 0)
+	// here. We don't use Difficulty  because CopyHeader can end up setting a
+	// nil Difficulty to a zero difficulty, so testing for nil difficulty is
```

Comment:
Why does setting a nil difficulty to a zero difficulty cause a problem? We are checking for a zero gaslimit, not a nil gaslimit, so I would think they would be the same right?

Reply:
Post gingerbread the difficulty is hardcoded to zero, so if we were to use `difficulty == 0` to select the encoding pre-gingerbread headers copied by `CopyHeader` would be encoded in a post gingerbread form.